src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl.xaml.cs

```diff
@@ -0,0 +1,388 @@
+                    if (caretPoint.HasValue)
+                    {
+                        var caretPosition = caretPoint.Value.Position;
+                        DocumentOutlineHelper.SelectDocumentNode(documentSymbolModelArray, activeTextView.TextSnapshot, LspSnapshot, caretPosition);
+                    }
+                }
+            }
+        }
+
+        /// <summary>
+        /// Starts a new task to get the current document model.
+        /// </summary>
+        private void StartGetModelTask()
+        {
+            _getModelQueue.AddWork();
+        }
+
+        /// <summary>
+        /// Starts a new task to update the current document model.
+        /// </summary>
+        private void StartModelUpdateTask()
+        {
+            if (LatestFetchedDocumentModelIsNotEmpty)
+                _updateModelQueue.AddWork();
+        }
+
+        /// <summary>
+        /// Starts a new task to highlight the symbol node corresponding to the current caret position in the editor.
+        /// </summary>
+        private void StartHightlightNodeTask()
+        {
+            if (LatestFetchedDocumentModelIsNotEmpty)
+                _highlightNodeQueue.AddWork();
+        }
+
+        private void ExpandAll(object sender, RoutedEventArgs e)
+        {
+            DocumentOutlineHelper.SetIsExpanded((IEnumerable<DocumentSymbolViewModel>)symbolTree.ItemsSource, true);
+        }
+
+        private void CollapseAll(object sender, RoutedEventArgs e)
+        {
+            DocumentOutlineHelper.SetIsExpanded((IEnumerable<DocumentSymbolViewModel>)symbolTree.ItemsSource, false);
+        }
+
+        private void Search(object sender, EventArgs e)
+        {
+            StartModelUpdateTask();
+        }
+
+        private void SortByName(object sender, EventArgs e)
+        {
+            SortOption = SortOption.Name;
+            StartModelUpdateTask();
+        }
+
+        private void SortByOrder(object sender, EventArgs e)
+        {
+            SortOption = SortOption.Order;
+            StartModelUpdateTask();
+        }
+
+        private void SortByType(object sender, EventArgs e)
+        {
+            SortOption = SortOption.Type;
+            StartModelUpdateTask();
+        }
+
+        // When symbol node clicked, select the corresponding code
+        private void JumpToContent(object sender, EventArgs e)
+        {
+            if (LatestFetchedDocumentModelIsNotEmpty)
+            {
+                CodeWindow.GetLastActiveView(out var textView);
+                var activeTextView = EditorAdaptersFactoryService.GetWpfTextView(textView);
+                if (activeTextView is not null && sender is StackPanel panel && panel.DataContext is DocumentSymbolViewModel symbol)
+                {
+                    // Avoids highlighting the node after moving the caret ourselves 
+                    // (The node is already highlighted on user click)
```

Comment:
Or is the goal here that we want to keep the original node highlighted even when we're clicking around?

Reply:
Yes, this is an optimization so that we don't kick off an extra task.

When the user clicks on a symbol node in the window, we want to move the cursor to that line. Without unsubscribing from `Caret_PositionChanged` first, we would call `StartHightlightNodeTask()` when we move the cursor ourselves. This is not ideal because it would be highlighting a node that's already highlighted. 

That's why after we move the caret, we resubscribe again so that when the user clicks somewhere else, we can highlight the new node.

Reply:
I added comments to clarify this :)